curl/src/client.cc

```diff
@@ -173,6 +171,79 @@ namespace leatherman { namespace curl {
 
+    void client::download_file(request const& req, std::string const& file_path)
+    {
+                   printf("Entered download_file\n");
+        response res;
+        context ctx(req, res);
+
+        // Reset the options
+        curl_easy_reset(_handle);
+
+        CURLcode result = CURLE_OK;
+        char errbuf[CURL_ERROR_SIZE] = { '\0' };
+        FILE* fp = nullptr;
+        try {
+            auto temp_path = fs::path(file_path).parent_path() / fs::unique_path("temp_file_%%%%-%%%%-%%%%-%%%%");
+            fp = boost::nowide::fopen(temp_path.string().c_str(), "wb");
+            if (!fp) {
+                   printf("About to throw exception b/c of fopen. This is to make sure CI passes\n");
+                throw http_file_download_exception(req, file_path, _("Failed to open temporary file for writing"));
+            }
+
+            curl_easy_setopt_maybe(ctx, CURLOPT_NOPROGRESS, 1);
+            curl_easy_setopt_maybe(ctx, CURLOPT_WRITEFUNCTION, write_file);
+            curl_easy_setopt_maybe(ctx, CURLOPT_WRITEDATA, fp);
+
+            // Setup the remaining request
+            set_url(ctx);
+            set_headers(ctx);
+            set_timeouts(ctx);
+            set_ca_info(ctx);
+            set_client_info(ctx);
+            set_client_protocols(ctx);
+
+            // More detailed error messages
+            curl_easy_setopt_maybe(ctx, CURLOPT_ERRORBUFFER, errbuf);
+
+                   printf("About to perform the request\n");
+            // Perform the request
+            result = curl_easy_perform(_handle);
+            if (result != CURLE_OK) {
+                    printf("Download failed, about to throw an exception\n");
+                fclose(fp);
+                fs::remove(temp_path);
+                throw http_file_download_exception(req, file_path, errbuf);
+            }
+            fclose(fp);
+
+            LOG_DEBUG("download completed, now writing result to file");
+            fs::rename(temp_path, file_path);
+        } catch (http_file_download_exception& e) {
+                 printf("Threw a file_download exception, re-throwing it.\n");
+            // so that the code in the lower catch block is not hit
+            throw e;
+        } catch (http_request_exception& e) {
+                 printf("Threw a request exception, converting to download\n");
+            fclose(fp);
+            throw http_file_download_exception(e.req(), file_path, e.what());
+        } catch (fs::filesystem_error& e) {
+                 printf("Threw a filesystem error, converting to download\n");
+            if (result == CURLE_OK) {
+                  printf("Error was before download.\n");
+              throw http_file_download_exception(req, file_path, e.what());
+            }
+                 printf("Error was in temp file removal\n"); 
+
+            throw http_file_download_exception(
+                req,
+                file_path,
+                errbuf + _(" and failed to remove temporary file"));
```

Comment:
This should use string substitution if we want it to translate usefully. As in `_("{1} and failed to remove temporary file", errbuf)`.